### PR TITLE
feat: Export insights table as CSV

### PR DESCRIPTION
**Pull request by timgl**
## Problem

As a user i want to be able to export the insights table as csv to do further analysis on

## Changes

- add trend.csv endpoint to export as csv
- add button to frontend

image
👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

unit tests
create insight, export as csv

### CODE REVIEW

**Review by clarkus — CHANGES_REQUESTED** [review]
The placement of the export action is problematic as it potentially interferes with placement for bulk actions. Bulk actions don't exist in this context, but as a system pattern I think we should reserve this area for those actions. The export action could also be more discoverable if we included a text label for the action. 

I think you could follow this pattern from the recent taxonomy and events work. The export action is placed above the table, in a secondary position with a clear label. You might need to include a header for the table just to balance that placement of the export icon. 

Screen Shot 2022-03-14 at 2 23 36 PM

File: posthog/api/insight.py [hunk 1]
```diff
@@ -303,6 +309,32 @@ def trend(self, request: request.Request, *args: Any, **kwargs: Any) -> Response
...
+            renderer = csvrenderers.CSVRenderer()
+            renderer.header = csvexport[0].keys()
+            export = renderer.render(csvexport)
+            if request.GET.get("export_insight_id"):
+                export = "{}/insights/{}/\n".format(SITE_URL, request.GET["export_insight_id"]).encode() + export
+
+            response = HttpResponse(export)
+            response[
+                "Content-Disposition"
+            ] = 'attachment; filename="posthog - {name} ({date_from} - {date_to}).csv"'.format(
```
Comment (Twixes) [inline posthog/api/insight.py:332]: Proposal for a subjectively slightly nicer/more helpful filename (putting "PostHog" at the end):
```suggestion
            ] = 'attachment; filename="{name} ({date_from}-{date_to}) from PostHog.csv"'.format(
```

**Review by Twixes — CHANGES_REQUESTED** [review]
This feature definitely makes sense, but I do agree with Chris that the placement of the button isn't great.

File: posthog/api/insight.py [hunk 2]
```diff
@@ -303,6 +309,32 @@ def trend(self, request: request.Request, *args: Any, **kwargs: Any) -> Response
...
+                    line[item["labels"][index]] = data
+                    total_sum += data
+
+                line["total sum"] = total_sum
+                csvexport.append(line)
+            renderer = csvrenderers.CSVRenderer()
+            renderer.header = csvexport[0].keys()
+            export = renderer.render(csvexport)
+            if request.GET.get("export_insight_id"):
+                export = "{}/insights/{}/\n".format(SITE_URL, request.GET["export_insight_id"]).encode() + export
```
[the remark this answers is not in the document]
Comment (timgl) [reply posthog/api/insight.py:327]: Hm yeah I think that's fair. Our friends at Ampl. include this in their export which i thought was smart as it makes it really easy to find where the data came from. Maybe the answer is to put it as the last column name?

**Comment by timgl** [comment]
Thanks both! Think the export button looks okay without a header in there but happy to add one if you think it needs it?
image

**Review by clarkus — APPROVED** [review]
:shipit: Don't think we need a header to ship this. Longer term I'd like to expand on the table patterns to include some proper reusable filtering and exporting controls. Was talking to @alexkim205 about this earlier since he did some of this work recently for data management / live events. It'd just be nice to have some reusable components that always work with card or list views.

**Comment by alexkim205** [comment]
> It'd just be nice to have some reusable components that always work with card or list views.

Made an issue here https://github.com/PostHog/posthog/issues/9035 to capture this initiative, but non blocking for this PR.

**Pull request merged**